### ISSUE TITLE
Incorrect height if only viewBox is specified in SVG

### ISSUE THREAD

**Issue by mojavelinux** [mogest/prawn-svg#55]
If only the viewBox is specified in the SVG, but no explicit width and height, Prawn SVG is not calculating the output height correctly (or, at least, it's not doing what I would expect). Instead of scaling the height read from the viewBox by the same amount as the width, it is scaling the available height on the page. As a result, a gap appears between the image and the next content.

Consider the following SVG (a red square):

```xml
<svg
  xmlns="http://www.w3.org/2000/svg"
  version="1.1"
  width="100" height="100"
  viewBox="0 0 100 100">
 <g>
  <rect style="fill:#800000" height="100" width="100" y="0" x="0"/>
 </g>
</svg>
```

If I embed this SVG using an explicit width of 250:

```ruby
Prawn::Document.generate 'svg-test.pdf' do
  svg ::IO.read('red-square.svg'), at: [0, cursor], width: 250
  text 'text under image'
end
```

The text appears below the square directly adjacent to it.

> ![svg-test-1](https://cloud.githubusercontent.com/assets/79351/8048890/4b913378-0e14-11e5-9219-f1874185dc54.png)

If I remove the width and height attributes from the SVG:

```xml
<svg
  xmlns="http://www.w3.org/2000/svg"
  version="1.1"
  viewBox="0 0 100 100">
 <g>
  <rect style="fill:#800000" height="100" width="100" y="0" x="0"/>
 </g>
</svg>
```

then embed it using the same Ruby code, then a gap appears under the image.

> ![svg-test-2](https://cloud.githubusercontent.com/assets/79351/8048913/940ddfd4-0e14-11e5-9cab-4f228566c5eb.png)

I think the problem is the default value for the height (and width). The height should fallback to the viewBox height (if specified) instead of the bounds height.

**Comment by mogest** [comment]
After experimenting a bit, I think this approach wasn't the right one.

From http://www.w3.org/TR/SVG/struct.html#SVGElementWidthAttribute:
> If the attribute is not specified, the effect is as if a value of '100%' were specified.

It's the same for height.  I'll make that change now.

**Comment by mogest** [comment]
This "fix" actually makes your issue above a lot worse (!), as it centres the unspecified width/height SVG in the middle of the page.  But I think if you don't specify a :height to the Ruby svg call, and you don't specify a height in the SVG, the spec is pretty clear that it should assume height="100%", and the only thing we have to go on to measure against is the bounds height.  Now, at least, it performs as Chrome and Safari do!

The viewbox width+height is only used to transform the real width and height; the document is still output to the size of the real width and height, but transformed so it uses the viewbox's coordinate system.  Taking the height from the viewbox when the document height hasn't been supplied is definitely the wrong approach, and makes documents render inconsistently with Chrome/Safari.

Good news though, is that going through this process found a really bad bug in the viewBox offset calculation!

So anyway, I think the solution here is for the user to specify the :width and :height to the Ruby svg call, in which case all works now as you'd expect it to.  What do you reckon?

**Comment by mojavelinux** [comment]
I agree with some of your conclusions and disagree with some others. I'll explain.

You are totally right that if a width is not specified in the image, and not passed to the `svg` method, then the width should equal the bounds width (100%). I'm in 100% agreement there.

With that said, the height still needs to be calculated in a way that respects the aspect ratio of the image. That aspect ratio, as far as I understand it, is established by the viewBox.

We can't just set the width and height to the bounds width and height without taking into consideration the aspect ratio of the image. If we do that, we get a gap under the image, which is not what we want nor is it the behavior observed in the browser.

> it centres the unspecified width/height SVG in the middle of the page.

That's not the behavior I'm seeing.

> Taking the height from the viewbox when the document height hasn't been supplied is definitely the wrong approach, and makes documents render inconsistently with Chrome/Safari.

True. But it should be increase to the available height in a way that preserves the aspect ratio.

I'll give a more concrete example to help with the thought process.

If the viewBox is "0 0 100 100" and the width and height are unspecified, then the output width and the height should both be equal to the bounds width. It would be incorrect if the output width equaled the bounds width and the output height equaled the bounds height. Then, there would be more space reserved for the SVG canvas than drawables and the user would observe a gap under the image. Browsers don't behave this way as far as I can tell.

In other words, if there is a viewBox, the calculations should be:

```ruby
width ||= @bounds[0]
height ||= ((width * @viewport_height) / @viewport_width.to_f)
```

Of course, that assumes we are maintaining the aspect ratio, so that probably needs to be taken into consideration. The default is to preserve it afaik.

There's also a problem with the following calculation because it possibly divides two integers, which can result in a 0 value for numbers lass than 1. 

```
scale = @requested_width / width
```

It should be:

```
scale = @requested_width / width.to_f
```

(or we need to ensure that the values are already floats somewhere else).

**Comment by mojavelinux** [comment]
One way to verify that we have the right behavior is to check that square example I cited and make sure we aren't getting a gap under the image. If we have a gap under a square image, we are doing something wrong ;)

**Comment by mogest** [comment]
I've just done some experimentation using Chrome with SVG inside HTML, as I guess that's the closest thing we have to go off, and the specs don't enlighten me as to what the right thing is to do.

Interestingly, if you specify width="100%" height="100%" on the svg tag, it behaves differently than as opposed to leaving them out altogether!  If you just render the SVG, without it being in HTML, it renders as height="100%".

Like you suggest, inside HTML when no height is supplied, the ratio of the viewBox seems to determine the height, the width being always held at 100% whether it's supplied or not.

I'll work at implementing that now!

**Comment by mogest** [comment]
OK, how does this seem to you?

With your example code, you'll get a 250x250 sized box, and the text will be directly underneath!  If you take out the :width parameter to the svg call, it makes a 540x540 sized box, and yep, text appears underneath.

If you take out the viewBox, well, you get a 100x100 box in a 300x150 space.  I think at this stage, you are just being a mean developer and you should give prawn-svg some hint about what you want :)

**Comment by mojavelinux** [comment]
Sweet!!! That is rocking! I can't wait to upgrade Asciidoctor PDF to use it. This is definitely making our next release.

> If you take out the viewBox, well, you get a 100x100 box in a 300x150 space. I think at this stage, you are just being a mean developer and you should give prawn-svg some hint about what you want :)

I totally agree. I don't even know what an SVG is without a viewBox. I think most SVG designers would consider that to be invalid syntax.

**Comment by mogest** [comment]
Thanks for your help!

Released in version 0.20.0.

**Comment by mojavelinux** [comment]
Excellent! I'll upgrade Asciidoctor PDF tonight. Woot!

**Comment by mojavelinux** [comment]
Updated! Also, thanks for creating the tag.
